util/topk/topk.go

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package topk defines a cheap probabilistic top-K data structure.
+package topk
+
+import (
+	"container/heap"
+	"hash/maphash"
+	"math"
+	"sync"
+)
+
+// TopK is a probabilistic counter of the top K items, using a count-min sketch
+// to keep track of item counts and a heap to track the top K of them.
+type TopK[T any] struct {
+	heap minHeap[T]
+	k    int
+	sf   SerializeFunc[T]
+	cms  CountMinSketch
+}
+
+// HashFunc is responsible for providing a []byte serialization of a value,
+// appended to the provided byte slice. This is used for hashing the value when
+// adding to a CountMinSketch.
+type SerializeFunc[T any] func([]byte, T) []byte
+
+// New creates a new TopK that stores k values. Parameters for the underlying
+// count-min sketch are chosen for a 0.1% error rate and a 0.1% probability of
+// error.
+func New[T any](k int, sf SerializeFunc[T]) *TopK[T] {
+	hashes, buckets := PickParams(0.001, 0.001)
+	return NewWithParams(k, sf, hashes, buckets)
+}
+
+// NewWithParams creates a new TopK that stores k values, and additionally
+// allows customizing the parameters for the underlying count-min sketch.
+func NewWithParams[T any](k int, sf SerializeFunc[T], numHashes, numCols int) *TopK[T] {
+	ret := &TopK[T]{
+		heap: make(minHeap[T], 0, k),
+		k:    k,
+		sf:   sf,
+	}
+	ret.cms.init(numHashes, numCols)
+	return ret
+}
+
+// Add calls AddN(val, 1).
+func (tk *TopK[T]) Add(val T) uint64 {
+	return tk.AddN(val, 1)
+}
+
+var hashPool = &sync.Pool{
+	New: func() any {
+		buf := make([]byte, 0, 128)
+		return &buf
+	},
+}
+
+// AddN adds the given item to the set with the provided count, returning the
+// new estimated count.
+func (tk *TopK[T]) AddN(val T, count uint64) uint64 {
+	buf := hashPool.Get().(*[]byte)
+	defer hashPool.Put(buf)
+	ser := tk.sf((*buf)[:0], val)
+
+	vcount := tk.cms.AddN(ser, count)
+
+	// If we don't have a full heap, just push it.
+	if len(tk.heap) < tk.k {
+		heap.Push(&tk.heap, mhValue[T]{
+			count: vcount,
+			val:   val,
+		})
+		return vcount
+	}
+
+	// If this item's count surpasses the heap's minimum, update the heap.
+	if vcount > tk.heap[0].count {
+		tk.heap[0] = mhValue[T]{
+			count: vcount,
+			val:   val,
+		}
+		heap.Fix(&tk.heap, 0)
+	}
+	return vcount
+}
+
+// Top returns the estimated top K items as stored by this TopK.
+func (tk *TopK[T]) Top() []T {
```

Comment:
AppendTop variant to let user supply the 0-len non-zero-cap slice to avoid allocs?